applications/nrf_desktop/src/modules/watchdog.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <device.h>
+#include <drivers/watchdog.h>
+
+#define MODULE watchdog
+#include "module_state_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_WATCHDOG_LOG_LEVEL);
+
+
+#if defined(CONFIG_WDT_0_NAME)
+#define WATCHDOG_DEV_NAME CONFIG_WDT_0_NAME
+#elif defined(DT_WDT_0_NAME)
+#define WATCHDOG_DEV_NAME DT_WDT_0_NAME
+#else
+#error "Error!!! Watchdog device name not defined - verify configuration."
+#endif
+#define WDT_FEED_WORKER_DELAY_MS 2000
+#define WDT_FEED_WORKER_START_IMMEDIATELY 0
+
+#if defined(CONFIG_DESKTOP_WATCHDOG_TIMEOUT)
+#define WDT_TIMEOUT_MS K_SECONDS(CONFIG_DESKTOP_WATCHDOG_TIMEOUT)
+#else
+/* 16s - required to survive LOG_PANIC print when power down */
+#define WDT_TIMEOUT_MS 16000
+#endif
+
+struct wdt_data_storage {
+	struct device *wdt_drv;
+	int wdt_channel_id;
+	struct k_delayed_work work;
+};
+static struct wdt_data_storage wdt_data;
+
+void watchdog_feed_worker(struct k_work *work_desc)
+{
+	struct wdt_data_storage *data =
+			CONTAINER_OF(work_desc, struct wdt_data_storage, work);
+
+	int err = wdt_feed(data->wdt_drv, data->wdt_channel_id);
+
+	if (err < 0) {
+		LOG_ERR("Cannot feed watchdog. Error code: %d", err);
+		module_set_state(MODULE_STATE_ERROR);
+		return;
+	}
+
+	k_delayed_work_submit(&data->work, WDT_FEED_WORKER_DELAY_MS);
+}
+
+void wdt_timeout_callback(struct device *dev, int channel_id)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(channel_id);
+	LOG_ERR("Watchdog timeout elapsed. Restarting the device!");
```

Comment:
We may not see this log (reboot may occur before processing it).

Reply:
If that is a problem then it will be changed.